model/ids_test.go

```diff
@@ -20,7 +20,7 @@ import (
 	"testing"
 
 	"github.com/gogo/protobuf/jsonpb"
-	"github.com/golang/protobuf/proto"
+	"github.com/gogo/protobuf/proto"
```

Comment:
@yurishkuro I could not make this test to work with the `"github.com/golang/protobuf/proto"`. 

Maybe the gogo is not fully compatible with protobuf 1.4.2? protobuf 1.4.2 should be backed by the V2 `google.golang.org/protobuf` https://blog.golang.org/protobuf-apiv2.

Reply:
Don't know, did you see if gogo project has any issues about that?

The import change looks suspicious, it changes what the test does, even different marshal functions would be used. But not sure if that's necessarily bad as long as other tests work, where the full trace is unmarshalled.

However, most of the tests only marshall JSON rep, not the binary format.

Reply:
I could not find anything exactly related to this https://github.com/gogo/protobuf/issues?q=is%3Aissue+is%3Aopen++invalid+Go+type

When using the original package `google.golang.org/protobuf` it results in:

```
panic: invalid Go type model.TraceID for field jaeger.api_v2.SpanRef.trace_id [recovered]
	panic: invalid Go type model.TraceID for field jaeger.api_v2.SpanRef.trace_id

goroutine 25 [running]:
testing.tRunner.func1.1(0x96dc60, 0xc0001a0eb0)
	/home/ploffay/bin/go/src/testing/testing.go:940 +0x2f5
testing.tRunner.func1(0xc00014bd40)
	/home/ploffay/bin/go/src/testing/testing.go:943 +0x3f9
panic(0x96dc60, 0xc0001a0eb0)
	/home/ploffay/bin/go/src/runtime/panic.go:969 +0x166
google.golang.org/protobuf/internal/impl.newSingularConverter(0xb04660, 0x9e3c20, 0xb04540, 0xc0001c8840, 0xc0001bc270, 0x30)
	/home/ploffay/projects/golang/pkg/mod/google.golang.org/protobuf@v1.25.0/internal/impl/convert.go:143 +0xff5
google.golang.org/protobuf/internal/impl.NewConverter(0xb04660, 0x9e3c20, 0xb04540, 0xc0001c8840, 0x2, 0x30)
	/home/ploffay/projects/golang/pkg/mod/google.golang.org/protobuf@v1.25.0/internal/impl/convert.go:60 +0xd9
google.golang.org/protobuf/internal/impl.fieldInfoForScalar(0xb04540, 0xc0001c8840, 0x99b44f, 0x7, 0x0, 0x0, 0xb04660, 0x9e3c20, 0x99b458, 0x5b, ...)
	/home/ploffay/projects/golang/pkg/mod/google.golang.org/protobuf@v1.25.0/internal/impl/message_reflect_field.go:234 +0x110
google.golang.org/protobuf/internal/impl.(*MessageInfo).makeKnownFieldsFunc(0xc000175040, 0x38, 0xffffffffffffffff, 0x20, 0xffffffffffffffff, 0xc000194c30, 0xc000194c60, 0xc000194c90, 0xc000194cc0)
	/home/ploffay/projects/golang/pkg/mod/google.golang.org/protobuf@v1.25.0/internal/impl/message_reflect.go:67 +0x938
google.golang.org/protobuf/internal/impl.(*MessageInfo).makeReflectFuncs(0xc000175040, 0xb04660, 0x9f4240, 0x38, 0xffffffffffffffff, 0x20, 0xffffffffffffffff, 0xc000194c30, 0xc000194c60, 0xc000194c90, ...)
	/home/ploffay/projects/golang/pkg/mod/google.golang.org/protobuf@v1.25.0/internal/impl/message_reflect.go:36 +0x63
google.golang.org/protobuf/internal/impl.(*MessageInfo).initOnce(0xc000175040)
	/home/ploffay/projects/golang/pkg/mod/google.golang.org/protobuf@v1.25.0/internal/impl/message.go:90 +0x182
google.golang.org/protobuf/internal/impl.(*MessageInfo).init(0xc000175040)
	/home/ploffay/projects/golang/pkg/mod/google.golang.org/protobuf@v1.25.0/internal/impl/message.go:72 +0x3c
google.golang.org/protobuf/internal/impl.(*messageReflectWrapper).ProtoMethods(0xc0001a0bb0, 0xc00005bc00)
	/home/ploffay/projects/golang/pkg/mod/google.golang.org/protobuf@v1.25.0/internal/impl/message_reflect_gen.go:150 +0x2f
google.golang.org/protobuf/proto.protoMethods(...)
	/home/ploffay/projects/golang/pkg/mod/google.golang.org/protobuf@v1.25.0/proto/proto_methods.go:18
google.golang.org/protobuf/proto.MarshalOptions.marshal(0xc000000001, 0x0, 0x0, 0x0, 0xb00fc0, 0xc0001a0bb0, 0xc0001a0bb0, 0xb00fc0, 0xc0001a0bb0, 0x9fe120, ...)
	/home/ploffay/projects/golang/pkg/mod/google.golang.org/protobuf@v1.25.0/proto/encode.go:143 +0x49
google.golang.org/protobuf/proto.MarshalOptions.MarshalAppend(0x1, 0x0, 0x0, 0x0, 0xaed960, 0xc0001a0bb0, 0x40ded6, 0xc00012a5a0, 0x60, 0x58, ...)
	/home/ploffay/projects/golang/pkg/mod/google.golang.org/protobuf@v1.25.0/proto/encode.go:125 +0x98
github.com/golang/protobuf/proto.marshalAppend(0x0, 0x0, 0x0, 0xaf84a0, 0xc000169e00, 0xc000048500, 0x40e748, 0x8, 0x978c80, 0x1, ...)
	/home/ploffay/projects/golang/pkg/mod/github.com/golang/protobuf@v1.4.2/proto/wire.go:40 +0xbe
github.com/golang/protobuf/proto.Marshal(0xaf84a0, 0xc000169e00, 0x0, 0x9d38e0, 0xa2cac5, 0x8, 0xd0)
	/home/ploffay/projects/golang/pkg/mod/github.com/golang/protobuf@v1.4.2/proto/wire.go:23 +0x4f
github.com/jaegertracing/jaeger/model_test.TestTraceSpanIDMarshalProto.func3(0xc00014bd40)
	/home/ploffay/projects/jaegertracing/jaeger/model/ids_test.go:74 +0x1be
testing.tRunner(0xc00014bd40, 0xc0001a0b70)
	/home/ploffay/bin/go/src/testing/testing.go:991 +0xdc
created by testing.(*T).Run
	/home/ploffay/bin/go/src/testing/testing.go:1042 +0x357

Process finished with exit code 1
```

Reply:
OTEL seems to be using only `gogo` imports https://github.com/open-telemetry/opentelemetry-collector/blob/aeb604f4a3227415ba6e209363fcaf7d93817b70/internal/data/opentelemetry-proto-gen/trace/v1/trace.pb.go#L13 I am not sure if we should do the same

Reply:
I am not opposed to import change, as long as things continue working. Although I am curious why I didn't use it in the first place.

```
panic: invalid Go type model.TraceID for field jaeger.api_v2.SpanRef.trace_id [recovered]
	panic: invalid Go type model.TraceID for field jaeger.api_v2.SpanRef.trace_id
```

If all the marshalling code is in the golang package, then it wouldn't know anything about `model.TraceID`, because that type is associated with the field only via gogo annotation. However, if you look at this stack in master, even though it starts with `golang.proto`, it soon calls into `(*SpanRef).XXX_Unmarshal`, which is auto-generated for our types. However, in your stack trace it continue using reflection all the way up.

```
github.com/jaegertracing/jaeger/model.(*TraceID).Unmarshal(...)
	/Users/ysh/dev/jaegertracing/jaeger/model/ids.go:121
github.com/jaegertracing/jaeger/model.(*SpanRef).Unmarshal(0xc000089e00, 0xc000028780, 0x1c, 0x1c, 0x1, 0x1fb5188)
	/Users/ysh/dev/jaegertracing/jaeger/model/model.pb.go:2086 +0x7b2
github.com/jaegertracing/jaeger/model.(*SpanRef).XXX_Unmarshal(0xc000089e00, 0xc000028780, 0x1c, 0x1c, 0xc000089e00, 0xc000072d01)
	/Users/ysh/dev/jaegertracing/jaeger/model/model.pb.go:258 +0x49
github.com/golang/protobuf/proto.Unmarshal(0xc000028780, 0x1c, 0x1c, 0x1646ac0, 0xc000089e00, 0xc00000f100, 0x0)
	/Users/ysh/golang/pkg/mod/github.com/golang/protobuf@v1.3.4/proto/decode.go:337 +0x1bc
github.com/jaegertracing/jaeger/model_test.TestTraceSpanIDMarshalProto.func3(0xc0000c19e0)
	/Users/ysh/dev/jaegertracing/jaeger/model/ids_test.go:83 +0x3c9
```

Reply:
What do you suggest then? I cannot say that I understand what is really wrong here.

Reply:
let's keep the new import